compiler/luci/service/src/Nodes/CircleReshape.cpp

```diff
@@ -34,4 +40,125 @@ luci::CircleNode *CloneNodeLet<CN::OPQR>::visit(const luci::CircleReshape *node)
   return cloned;
 }
 
+namespace sinf
+{
+
+/**
+ * @note  CircleReshape always has two inputs: tensor and shape.
+ *        The shape input can be CircleConst, CircleOutputDummy, or CircleNode.
+ *        - If the shape input is CircleConst, the shape is inferred from the constant.
+ *        - If the shape input is CircleOutputDummy, the shape is inferred from
+ *          the attribute or the node itself.
+ *        - If the shape input is CircleNode, the shape is not inferred.
+ */
+loco::TensorShape Algorithm::visit(const luci::CircleReshape *node)
+{
+  const loco::DataType S32 = loco::DataType::S32;
+
+  // CircleReshape node must have reshape/shape
+  if (node->shape() == nullptr)
+  {
+    INTERNAL_EXN("2nd input shape() should not be nullptr");
+  }
+
+  bool should_infer = true;
+  loco::TensorShape output_shape;
+  {
+    // Check if reshape/shape is CircleConst
+    auto const_input = dynamic_cast<luci::CircleConst *>(node->shape());
+    if (const_input != nullptr)
+    {
+      output_shape.rank(const_input->size<S32>());
+
+      for (uint32_t axis = 0; axis < output_shape.rank(); ++axis)
+      {
+        output_shape.dim(axis) = const_input->at<S32>(axis);
+      }
+    }
+    else
+    {
+      // Check if reshape/shape is CircleOutputDummy
+      auto dummy_input = dynamic_cast<luci::CircleOutputDummy *>(node->shape());
+      if (dummy_input != nullptr)
+      {
+        if (node->newShape()->rank() > 0)
+        {
+          output_shape.rank(node->newShape()->rank());
+
+          for (uint32_t axis = 0; axis < output_shape.rank(); ++axis)
+          {
+            output_shape.dim(axis) = node->newShape()->dim(axis);
+          }
+        }
+        else
+        {
+          output_shape = circle_shape(node);
+        }
+      }
+      else
+      {
+        // Check if reshape/shape is CircleNode
+        auto node_input = dynamic_cast<luci::CircleNode *>(node->shape());
+        if (node_input != nullptr)
+        {
+          output_shape.rank(node_input->dim(0).value());
+
+          for (uint32_t axis = 0; axis < output_shape.rank(); ++axis)
+          {
+            output_shape.dim(axis).unset();
+          }
+
+          should_infer = false;
+        }
+      }
+    }
+  }
+
+  const auto input = loco::must_cast<luci::CircleNode *>(node->tensor());
+  const auto input_shape = circle_shape(input);
+  uint32_t input_element_count = 1;
+  for (uint32_t axis = 0; axis < input_shape.rank(); ++axis)
+  {
+    if (input_shape.dim(axis).known())
+    {
+      input_element_count *= input_shape.dim(axis).value();
+    }
+    else
+    {
+      should_infer = false;
+      break;
+    }
+  }
+
+  if (should_infer)
+  {
+    uint32_t output_element_count = 1;
+    uint32_t unknown_dim_index = UINT32_MAX;
+    for (uint32_t dim_index = 0; dim_index < output_shape.rank(); ++dim_index)
+    {
+      const uint32_t dim_value = output_shape.dim(dim_index).value();
+      if (static_cast<int>(dim_value) == -1)
```

Comment:
This is not recommended code :(
`if (output_shape.dim(dim_index).known() == false)` would be better

Reply:
That is so true. Thank you!

I'll make the change.